helix-core/src/main/java/org/apache/helix/spectator/RoutingTableSnapshot.java

```diff
@@ -0,0 +1,151 @@
+package org.apache.helix.spectator;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.LiveInstance;
+
+/**
+ * The snapshot of RoutingTable information.  It is immutable, it reflects the routing table
+ * information at the time it is generated.
+ */
+public class RoutingTableSnapshot {
```

Comment:
whats the benefit of this wrapper on RoutingTable?

Reply:
The routingtable information in RoutingTableProvider keeps changing as long as there are externalview changes, however, we have a  use case that the application wants a consitent view of a whole routingtable (externalviews). Their code will iterate through the routingtable, if anything changes the table during their iteration, it breaks their code assumption.

Reply:
So we add this feature to return a snapshot of routingtable at that time.  The routingTable in the provider will keep update, but the snapshot is a read-only view of routingtable at the time it was created.